### ISSUE TITLE
[dart2wasm] Function equality is implemented as virtual call, but could be direct call

### ISSUE THREAD

**Issue by osa1** [dart-lang/sdk#51028]
```dart
void main() {
  X foo<X>(X x) => x;
  int Function(int) x1 = foo<int>;
  int Function(int) x2 = foo<int>;
  print(x1 == x2);
}
```

Here the equality is currently compiled to a virtual call, because `translator.singleTarget(node)` for the `EqualityNode` is returning `null`.

We have one `Function` implementation (the `_Function` class, which I think should be named `_Closure`), which overrides `operator ==`. So any `==` calls on `Function` can be a direct call to `_Function.==`. It seems like `DirectCallMetadataRepository` is not aware of this.

This kind of thing is done on other example, for example below the equality is direct call to `B.==`:

```dart
abstract class A {}

class B implements A {
  @override
  bool operator==(Object other) {
    return other is B;
  }
}

void main() {
  final A a = B();
  final A b = B();
  print(a == b);
}
```

**Comment by osa1** [comment]
> the _Function class, which I think should be named _Closure

Renaming done as a part of 5d77ea8cfe7abe3110198d5bf141276f89d9a425.

**Comment by osa1** [comment]
Fixed with 8fc7881fc87efd383b69c0aac8eb11dca311d55a.
